From 51d9ac2ab1ea56954091ff0818c781605264af3c Mon Sep 17 00:00:00 2001
From: Devarsh Thakkar <devarsh.thakkar@xilinx.com>
Date: Mon, 26 Aug 2019 09:42:00 -0700
Subject: [PATCH 0666/1851] v4l: xilinx: dma: Disable auto restart for low
 latency capture

commit a5334926b635b9b52912512ea46817d0f1032f35 from
https://github.com/Xilinx/linux-xlnx.git

Set default mode of operation instead of AUTO_RESTART for low latency
capture as AUTO_RESTART mode may take extra one frame delay
between programming and the actual writing data to memory.

As for the low latency capture we want to submit the buffer to consumer
when dma has just started processing without keeping any extra delay.

And using auto-restart mode without this delay is not recommended as
it may lead to frame tear due to next component trying to use the buffer
while dma is actually writing into it.

Signed-off-by: Devarsh Thakkar <devarsh.thakkar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-dma.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/media/platform/xilinx/xilinx-dma.c b/drivers/media/platform/xilinx/xilinx-dma.c
index 5882358aa53f..50a264d1be8c 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.c
+++ b/drivers/media/platform/xilinx/xilinx-dma.c
@@ -1148,11 +1148,18 @@ xvip_dma_set_ctrl(struct file *file, void *fh, struct v4l2_control *ctl)
 				return -EBUSY;
 
 			dma->low_latency_cap = true;
+			/*
+			 * Don't use auto-restart for low latency
+			 * to avoid extra one frame delay between
+			 * programming and actual writing of data
+			 */
+			xilinx_xdma_set_mode(dma->dma, DEFAULT);
 		} else if (ctl->value == XVIP_LOW_LATENCY_DISABLE) {
 			if (vb2_is_busy(&dma->queue))
 				return -EBUSY;
 
 			dma->low_latency_cap = false;
+			xilinx_xdma_set_mode(dma->dma, AUTO_RESTART);
 		} else if (ctl->value == XVIP_START_DMA) {
 			/*
 			 * In low latency capture, the driver allows application
-- 
2.31.1

